File: app/src/main/java/org/anvei/novelreader/widget/read/ReadPage.java
package org.anvei.novelreader.widget.read;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.Nullable;

import org.anvei.novelreader.widget.read.page.Page;
import org.anvei.novelreader.widget.read.page.PageConfig;

public class ReadPage extends View {

    private String chapterTitle;
    private boolean isFirstPage = false;

    private Page page;
    private Bitmap bitmap;
    private boolean needRecreate = false;

    private PageConfig pageConfig;

    public ReadPage(Context context) {
        super(context);
    }

    public ReadPage(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public void setChapterTitle(String chapterTitle) {
        this.chapterTitle = chapterTitle;
    }

    public void setIsFirstPage(boolean firstPage) {
        isFirstPage = firstPage;
    }

    public boolean isFirstPage() {
        return isFirstPage;
    }

    public void setPage(Page page) {
        this.page = page;
        requestReCreate();
    }

    public void requestReCreate() {
        needRecreate = true;
        postInvalidate();
    }

    public PageConfig getPageConfig() {
        return pageConfig;
    }

    public void setPageConfig(PageConfig pageConfig) {
        this.pageConfig = pageConfig;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (page != null) {
            if (bitmap == null || needRecreate) {
                if (pageConfig.width == 0 || pageConfig.height == 0) {
                    pageConfig.width = getWidth();
                    pageConfig.height = getHeight();
                }
                bitmap = pageConfig.pageFactory.createPage(this.page);
                if (needRecreate) {
                    needRecreate = false;
                }
            }
            canvas.drawBitmap(bitmap, 0F, 0F, pageConfig.paint);
        }
    }
}

File: app/src/main/java/org/anvei/novelreader/widget/read/ReadPage.kt
package org.anvei.novelreader.widget.read

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import org.anvei.novelreader.R
import org.anvei.novelreader.widget.read.page.Page

class ReadPage(context: Context, attributeSet: AttributeSet?) : ViewGroup(context, attributeSet) {

    var headerView: View? = null            // 页眉
    var footerView: View? = null            // 页脚
    var titleView: View? = null             // 章节标题视图
    lateinit var contentView: ContentView           // 章节内容视图

    lateinit var view: View

    var title: String = ""                  // 章节标题
    var content: Page? = null               // 章节内容

    constructor(context: Context) : this(context, null)

    init {
        attributeSet?.let {
            val typedArray = context.obtainStyledAttributes(attributeSet, R.styleable.ReadPage)
            val id = typedArray.getResourceId(R.styleable.ReadPage_layout_child, -1)
            if (id != -1) {
                throw IllegalStateException("child属性是必须的！")
            }
            setView(id)
            typedArray.recycle()
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        for (i in 0 until childCount) {
            getChildAt(i).measure(widthMeasureSpec, heightMeasureSpec)
        }
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        for (i in 0 until childCount) {
            getChildAt(i).apply {
                layout(0, 0, this.measuredWidth, this.measuredHeight)
            }
        }
    }

    fun showTitleView(show: Boolean) {
        titleView?.visibility = if (show) VISIBLE else GONE
    }

    fun setView(view: Int) {
        this.view = LayoutInflater.from(context).inflate(view, this)
    }

    fun getTitleHeight(): Int {
        if (titleView != null) {
            return titleView!!.height
        }
        return 0;
    }
}